src/libraries/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.cs

```diff
@@ -70,16 +70,42 @@ internal static JsonValueKind ToValueKind(this JsonTokenType tokenType)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static int IndexOfQuoteOrAnyControlOrBackSlash(this ReadOnlySpan<byte> span)
         {
+            if (Vector.IsHardwareAccelerated)
+            {
+                // For performance, use unsafe pointers and Vector.
+                return IndexOfOrLessThan_Vector(
+                        ref MemoryMarshal.GetReference(span),
+                        JsonConstants.Quote,
+                        JsonConstants.BackSlash,
+                        lessThan: 32,   // Space ' '
+                        span.Length);
+            }
+
+            // For performance, avoid unsafe pointers.
             return IndexOfOrLessThan(
-                    ref MemoryMarshal.GetReference(span),
-                    JsonConstants.Quote,
-                    JsonConstants.BackSlash,
-                    lessThan: 32,   // Space ' '
-                    span.Length);
+                span,
+                JsonConstants.Quote,
+                JsonConstants.BackSlash,
+                lessThan: 32);   // Space ' '
         }
 
-        private static unsafe int IndexOfOrLessThan(ref byte searchSpace, byte value0, byte value1, byte lessThan, int length)
+        private static unsafe int IndexOfOrLessThan(ReadOnlySpan<byte> span, byte value0, byte value1, byte lessThan)
```

Comment:
What length is the break-even point where this is faster than the vectorized code below? 

I am surprised this provides wins over the loop-unrolled sequential scan. Where is the overhead coming from?

This code was inspired by the general-purpose IndexOf method on `Span<T>`, here:
https://github.com/dotnet/runtime/blob/d8cf13e0ba9b369a15a83472b6b97463c6d07fe2/src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs#L189

Ideally, we should fix-up and make **that method** close to fastest for all input lengths to avoid having two separate code-paths, since that is used in a lot more places. I am not sure it is worth it to make a change like this here, just specific to JSON reading.

Reply:
> What length is the break-even point where this is faster than the vectorized code below?

The manual loop is always faster when running under the Mono interpreter (I tried from 10 characters to 10,240 characters as per the table above).

It would be good to know when else `Vector.IsHardwareAccelerated` is false so I could run some benchmarks other than Mono interpreter..

Reply:
What about less than 10 characters? For example, it is common for property names to be between 3-6 chars.

> The manual loop is always faster when running under the Mono interpreter

What about the regular JIT/compiled environments? I am concerned this will hurt those environments, particularly for larger strings which could benefit from loop unrolling and vectorizing.

Reply:
I want to echo what @jkotas mentioned here: https://github.com/dotnet/runtime/pull/39323#discussion_r455089989

We should try to avoid trade-offs that try to improve the interpreter environment at the cost of JIT environments and code maintainability/duplication. 

However, if **all cases** improve, then I can see an argument to be made for such a change.

Reply:
Run with `COMPlus_EnableHWIntrinsic=0` to disable hardware intrinsics.

Reply:
Is this a place where Mono could improve their SIMD/Vector<T>/HWIntrinsic support for bigger wins?

Reply:
These optimizations are focused on improving the WASM interpreter which doesn't have SIMD/Vectors in MVP platform. The numbers appear to show that the JIT doesn't suffer noticably?

Reply:
The existing IndexOfOrLessThan method should work just fine on platforms without SIMD/Vectors. `Vector.IsHardwareAccelerated` will be constant and then it will go into manually unrolled non-SIMD loop.

I believe that this change will make the performance measurably worse on platforms without SIMD/Vector acceleration (e.g. CoreCLR arm) because we will take this version that does not have the loop unrolled.

Reply:
There may also be SIMD support for WASM eventually. My understanding is that Google already supports it in V8 and there is an entire repo + proposals/extensions attempting to officially add SIMD as an available extension: https://github.com/WebAssembly/simd/issues/31 and https://webassembly.org/roadmap/

Reply:
> What about less than 10 characters

The perf is not affected when <= 10 chars under the interpreter (within the margin of error).

I increased the iterations to 100_000 with string length of 3:
```
Before
serialize:Serialization took 1536 ms
deserialize:Deserialization took 2141 ms

After
serialize:Serialization took 1550 ms
deserialize:Deserialization took 2072 ms
```

Reply:
> These optimizations are focused on improving the WASM interpreter which doesn't have SIMD/Vectors in the MVP platform. The numbers appear to show that the JIT doesn't suffer measurably?

This PR was intended to help with Mono interpreter, but helps even more with Mono JIT. However, the original code path remains for Core JIT unless `Vector.IsHardwareAccelerated==false` which means there is no hardware-accelerated SIMD support as is the case today with the Mono interpreter (whether or not running on WASM).

Reply:
> What about the regular JIT/compiled environments? I am concerned this will hurt those environments, particularly for larger strings which could benefit from loop unrolling.

Regular JIT/compiled environment use the existing code path unless `Vector.IsHardwareAccelerated==false`.

> I believe that this change will make the performance measurably worse on platforms without SIMD/Vector acceleration (e.g. CoreCLR arm) because we will take this version that does not have the loop unrolled.

Thanks for the CoreCLR arm reference. As I mentioned previously, I wasn't sure what other platforms `Vector.IsHardwareAccelerated` would be false.

I set `COMPlus_EnableHWIntrinsic=0` and re-ran the benchmarks under Core JIT. It was ~2x slower, so yes we should do something. I believe the unsafe code in the loop unrolling is causing the interpreter to be slow. There is currently not an option to detect the interpreter, and probably not something we want to expose anyway.

However, there is `RuntimeFeature.IsDynamicCodeCompiled` which will be false under the interpreter and could be used in conjunction with Vector.IsHardwareAccelerated` to ensure the manual loop that was added in this PR is only run when both are false. Thoughts?

Reply:
> What about the regular JIT/compiled environments? I am concerned this will hurt those environments, particularly for larger strings which could benefit from loop unrolling.

Regular JIT/compiled environment use the existing code path unless Vector.IsHardwareAccelerated==false.

> I believe that this change will make the performance measurably worse on platforms without SIMD/Vector acceleration (e.g. CoreCLR arm) because we will take this version that does not have the loop unrolled.

Thanks for the CoreCLR arm reference. As I mentioned previously, I wasn't sure what other platforms Vector.IsHardwareAccelerated would be false.

For a manual test, I commented out the first branch in `IndexOfOrLessThan(ReadOnlySpan<byte> span`
```cs
            if (length >= Vector<byte>.Count * 2)
            {
                int unaligned = (int)Unsafe.AsPointer(ref searchSpace) & (Vector<byte>.Count - 1);
                nLength = (IntPtr)((Vector<byte>.Count - unaligned) & (Vector<byte>.Count - 1));
            }
```
to avoid the Vector optimizations. The large 10K string test ran ~1.75x slower, so yes we should do something. I believe the unsafe code in the loop unrolling is causing the interpreter to be slow. There is currently not an option to detect the interpreter, and probably not something we want to expose anyway.

However, there is `RuntimeFeature.IsDynamicCodeCompiled` which will be false under the interpreter and could be used in conjunction with `Vector.IsHardwareAccelerated` to ensure the manual loop that was added in this PR is only run when both are false. Thoughts?

Reply:
Also, one side-affect of using `RuntimeFeature.IsDynamicCodeSupported` to detect the interpreter is that we will need to add a netstandard 2.1 assembly to the S.T.J package, since that API was added in 2.1 but we still support 2.0.